README.md
wptool
======

A command-line tool for looking at and fixing Wordpress installs.

Usage:

      wptool [dir] [options]
    
      Options:
        --check        Check MD5Sums
        --diff [path]  Show a diff of all changed files or, optionally, of the 
                       file at [path] relative to the dir.
        --fix          Check MD5Sums and replace changed files with new ones
        --help         See this help
        --mysql        Start a MySQL shell using credentials in wp-config.php
        --perms        Guess at sensible fs ownership and permissions; prompt 
                       to set them
        --user [name]  Override the unix user guessed by --perms
        --version      Print Wordpress version
        -V             Print the version of wptool


The options are more like (and should probably have been implemented as) commands. This 
tool uses a repository of Wordpress installs and associated MD5sum text files that can
be found here:

    http://wpsums.avi.co/

You can use a different URL by setting the evironment variable `WPSUMS_WORDPRESSES_URL`; 
that should be a URL to a directory containing a series of subdirectories each named
for a specific version; http://wpsums.avi.co/wordpresses/ is the default.

check
-----

Reads version.php, downloads the appropriate md5sums.txt file and checks each 
file in the instal has the 'correct' md5sum. The names of any that don't match 
are printed to stdout:

    avi@amazing:~$ wptool /home/avi/web/wordpress/public_html/ --check 
    changed: index.php
    changed: wp-includes/feed.php
    missing: wp-cron.php
    avi@amazing:~$ 


diff
----

Does a check and then prints a diff for each changed file:

    avi@amazing:~$ wptool /home/avi/web/wordpress/public_html/ --diff
    # < = Lines removed
    # > = Lines added
    index.php:
    1a2
    > include(/tmp/naughtythings.php)
    wp-includes/feed.php:
    179c179,183
    < 	echo apply_filters('the_excerpt_rss', $output);
    ---
    >		echo get_the_content_feed($feed_type);
    > 	}else{
    > 		echo apply_filters('the_excerpt_rss', $output);
    > 	}
    avi@amazing:~$ 

fix
---

Does a check and downloads new files to replace any changed or absent ones; 
no backups are taken:

    avi@amazing:~$ wptool /home/avi/web/wordpress/public_html/ --fix
    Changed: wp-includes/feed.php replaced
    Changed: index.php replaced
    Missing: wp-cron.php replaced
    avi@amazing:~$

mysql
-----

Parses wp_config.php and opens a MySQL shell using the credentials contained
therein:

    avi@amazing:~$ wptool /home/avi/web/wordpress/public_html/ --mysql
    Reading table information for completion of table and column names
    You can turn off this feature to get a quicker startup with -A
    
    Welcome to the MariaDB monitor.  Commands end with ; or \g.
    Your MariaDB connection id is 104
    Server version: 5.5.34-MariaDB-1~wheezy-log mariadb.org binary distribution
    
    Copyright (c) 2000, 2013, Oracle, Monty Program Ab and others.
    
    Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.
    
    MariaDB [wordpress]> 


perms
-----

Finds the owner of the directory passed and prompts to run these commands on it:

    chown <user>:www-data <dir> -R
    find <dir> -type f -exec chmod 740 {} \;
    find <dir> -type d -exec chmod 750 {} \;
	  chmod g+rws,u+s <dir>/wp-content/uploads -R

And, if <dir>/cgi-bin is found:

		chown <user>:<user> <dir>/cgi-bin/ -R
		chmod 755 <dir>/cgi-bin/ -R

The username may be overriden with the `--user` switch; you are prompted
with the suggested commands before they are run.


version
-------

Prints the version string as found in version.php:

    avi@amazing:~$ wptool /home/avi/web/wordpress/public_html/ --version
    3.8.1
		avi@amazing:~$



changes
0.20140217


lib/Wordpress.pm
#! /usr/bin/perl

package Wordpress;

use LWP::Simple; 
use Digest::MD5 qw/md5_hex/;
use Data::Dumper;
use File::Temp qw/tempfile/;
use Term::ANSIColor;

use strict;
use Exporter;
use vars qw($VERSION @ISA @EXPORT @EXPORT_OK %EXPORT_TAGS);

@ISA    = qw/Exporter/;
@EXPORT = qw/documentRoot checkFiles fullPath catFile cleanFile sqlShellCmd setVersion installWordpress diffFile getMd5Sums/;

use constant PHP => "/usr/bin/php";
use constant WPSUMS_WORDPRESSES_URL => $ENV{'WPSUMS_WORDPRESSES_URL'} || "http://wpsums.avi.co/wordpresses/";
use constant MD5SUMS_FILE => "md5sums.txt";
use constant DIFF => "/usr/bin/diff";
use constant DIFF_OPTS => "";

my $wpFiles = {
	version => "wp-includes/version.php",
	config  => "wp-config.php",
};
my $versions = {} ;
my $wp_version;
my $dir;

# Running getVersion sets the $versions and $dir variables
# for the module.
sub documentRoot{
	$dir = shift;
	my $file = $dir."/".$wpFiles->{'version'};
	open(my $fh, "<", $file) or die ("Error opening WP version file $file");
	while(my $line = readline($fh)){
		if($line =~ /^\s*\$(\S+)\s*=\s*('|")?([\d\.]+)('|")?;\s*$/g){
			$versions->{$1} = $3;
		}
	}
	$wp_version = $versions->{'wp_version'};
	return $versions;
}
sub setVersion{
	$wp_version = shift;
	return $wp_version;
}

sub getMd5Sums{
	my $url = _getUrl(MD5SUMS_FILE);
	die('$wp_version not set') unless $wp_version =~ /.+/;
	my $sums = {};
	my $list = get($url);
	foreach my $line (split(/\n+/, $list)){
		chomp $line;
		my($hash,$path) = split(/\s+\.\//, $line);
		$sums->{$path} = $hash;
	}
	return $sums;
}

sub checkFiles{
	my $dir = shift;
	my $sums = getMd5Sums();
	my $differences = {};
	my $absences;
	foreach my $file (keys(%{$sums})){
		my $sum = $sums->{$file};
		my $file_absolute = fullPath($file);
		if(! -f $file_absolute){
			$absences->{$file}++;
			next;
		}
		my @sums = fileHasChanged($file_absolute, $sum);
		if($sums[1]){
			$differences->{$file} = \@sums;
		}
	}
	return $differences, $absences;
}

# this doesn't work yet.
sub installWordpress{
	$dir = shift;
	my $sums = getMd5Sums();
	my $count = 0;
	foreach my $file(keys(%$sums)){
		$count++;
		my $fullPath = fullPath($file);
		my $url = _getUrl($file);
		print "$url -> $fullPath\n";
		my $status = getstore($url, $fullPath);
		if (is_error($status)){
			_warn("Failed to download '$url' ($status)");
		}
	}
	my($differences, $absences) = checkFiles($dir);
	if(defined($differences) || defined ($absences)){
		_error("Something went wrong. Try running a check on the dir");
	}
	return $count;
}

sub fileHasChanged{
	my $file = shift;
	my $checksum= shift;
	my $actualsum = md5sum($file);
	unless($checksum eq $actualsum){
		return $checksum, $actualsum;
	}
	return;
}

sub md5sum{
	my $file = shift;
#	open(my $fh, "<", $file) or return;
#	my $data = <$fh>;
#	close($fh);
#	return md5_hex($data);
	return (split(/\s+/, `md5sum $file`))[0];

}

sub _getUrl{
	my $file = shift;
	return WPSUMS_WORDPRESSES_URL.$wp_version."/".$file;
}


sub catFile{
	my $file = shift;
	open(my $fh, "<", $file) or return $!;
	my @file = <$fh>;
	close($fh);
	return join("", @file);
}

sub fullPath{
	my $relative = shift;
	my $fullPath = $dir."/".$relative;
	$fullPath =~ s#//#/#g;
	$fullPath =~ s#/./#/#;
	return $fullPath;
}

sub downloadFile{
	my $url = shift;
	my $path = shift;
	my $status = getstore($url, $path);
	if(is_error($status)){
		print STDERR "Download failed [$status] ($url)\n";
		return;
	}
	return $path;
}

sub diffFile{
	my $file = shift;
	my $fullPath = fullPath($file);
  my $url = _getUrl($file);
	my ($fh_tmp, $f_tmp) = tempfile();
	downloadFile($url, $f_tmp);
	my $cmd = join(" ", DIFF, DIFF_OPTS, $f_tmp, $fullPath);
	my @lines = `$cmd`;
	return @lines;
}

sub cleanFile{
#	my $dir = shift;
	my $file = shift;
	my $sum = shift;
	my $url = _getUrl($file);
	my $fullPath = fullPath($file);
	downloadFile($url, $fullPath);
	if(fileHasChanged($fullPath, $sum)){
		_warning("\nERROR: Attempted to replace '$fullPath'; failed $!");
		return;
	}
	return $sum;
}

sub sqlShellCmd{
	my $vars = parseVariables(fullPath("wp-config.php"));
	my $cmd = "mysql -u $vars->{'DB_USER'} -p$vars->{'DB_PASSWORD'} -h$vars->{'DB_HOST'} $vars->{'DB_NAME'}";
	return $cmd;
}


sub parseVariables{
	my $file = shift;
	open(my $fh, "<", $file) or die "Failed to open config.php : $!";
	my $vars;
	while(my $line = readline($fh)){
		#define lines:
		if($line =~ /^\s*define\s*\(["']([^"']+)["']\s*,\s*["']([^"']+)["']\s*\)\s*;\s*$/){
			$vars->{$1} = $2;
		# variable assignment:
		}elsif($line =~ /^\s*\$(\S+)\s*=\s*('|")?([\d\.]+)('|")?;\s*$/g){
			$vars->{$1} = $3;
		}
	}
	return $vars;
}


sub _error{
	my $message = shift;
	print STDERR $message, "\n";
	exit 1;
} 

sub _warning{
	my $message = shift;
	print STDERR $message, "\n";
}

1


lib/WpTool/Utilities.pm
#! /usr/bin/perl

package WpTool::Utilities;

use LWP::Simple; 
use Digest::MD5 qw/md5_hex/;
use Data::Dumper;
use File::Temp qw/tempfile/;
use Term::ANSIColor;

use strict;
use Exporter;
use vars qw($VERSION @ISA @EXPORT @EXPORT_OK %EXPORT_TAGS);

@ISA    = qw/Exporter/;
@EXPORT = qw/documentRoot checkFiles fullPath catFile cleanFile sqlShellCmd setVersion installWordpress diffFile getMd5Sums version getFixPermsCommands guessUnixUser/;

use constant PHP => "/usr/bin/php";
use constant WPSUMS_WORDPRESSES_URL => $ENV{'WPSUMS_WORDPRESSES_URL'} || "http://wpsums.avi.co/wordpresses/";
use constant MD5SUMS_FILE => "md5sums.txt";
use constant DIFF => "/usr/bin/diff";
use constant DIFF_OPTS => "";

our $VERSION=0.20140216;

my $wpFiles = {
	version => "wp-includes/version.php",
	config  => "wp-config.php",
};
my $versions = {} ;
my $wp_version;
my $dir;

# Running getVersion sets the $versions and $dir variables
# for the module.
sub documentRoot{
	$dir = shift;
	my $file = $dir."/".$wpFiles->{'version'};
	open(my $fh, "<", $file) or die ("Error opening WP version file $file");
	while(my $line = readline($fh)){
		if($line =~ /^\s*\$(\S+)\s*=\s*('|")?([\d\.]+)('|")?;\s*$/g){
			$versions->{$1} = $3;
		}
	}
	$wp_version = $versions->{'wp_version'};
	return $versions;
}
sub setVersion{
	$wp_version = shift;
	return $wp_version;
}

sub getMd5Sums{
	my $url = _getUrl(MD5SUMS_FILE);
	die('$wp_version not set') unless $wp_version =~ /.+/;
	my $sums = {};
	my $list = get($url);
	foreach my $line (split(/\n+/, $list)){
		chomp $line;
		my($hash,$path) = split(/\s+\.\//, $line);
		$sums->{$path} = $hash;
	}
	return $sums;
}

sub checkFiles{
	my $dir = shift;
	my $sums = getMd5Sums();
	my $differences = {};
	my $absences;
	foreach my $file (keys(%{$sums})){
		my $sum = $sums->{$file};
		my $file_absolute = fullPath($file);
		if(! -f $file_absolute){
			$absences->{$file}++;
			next;
		}
		my @sums = fileHasChanged($file_absolute, $sum);
		if($sums[1]){
			$differences->{$file} = \@sums;
		}
	}
	return $differences, $absences;
}

# this doesn't work yet.
sub installWordpress{
	$dir = shift;
	my $sums = getMd5Sums();
	my $count = 0;
	foreach my $file(keys(%$sums)){
		$count++;
		my $fullPath = fullPath($file);
		my $url = _getUrl($file);
		print "$url -> $fullPath\n";
		my $status = getstore($url, $fullPath);
		if (is_error($status)){
			_warn("Failed to download '$url' ($status)");
		}
	}
	my($differences, $absences) = checkFiles($dir);
	if(defined($differences) || defined ($absences)){
		_error("Something went wrong. Try running a check on the dir");
	}
	return $count;
}

sub fileHasChanged{
	my $file = shift;
	my $checksum= shift;
	my $actualsum = md5sum($file);
	unless($checksum eq $actualsum){
		return $checksum, $actualsum;
	}
	return;
}

sub md5sum{
	my $file = shift;
#	open(my $fh, "<", $file) or return;
#	my $data = <$fh>;
#	close($fh);
#	return md5_hex($data);
	return (split(/\s+/, `md5sum $file`))[0];

}

sub _getUrl{
	my $file = shift;
	return WPSUMS_WORDPRESSES_URL.$wp_version."/".$file;
}


sub catFile{
	my $file = shift;
	open(my $fh, "<", $file) or return $!;
	my @file = <$fh>;
	close($fh);
	return join("", @file);
}

sub fullPath{
	my $relative = shift;
	my $fullPath = $dir."/".$relative;
	$fullPath =~ s#//#/#g;
	$fullPath =~ s#/./#/#;
	return $fullPath;
}

sub downloadFile{
	my $url = shift;
	my $path = shift;
	my $status = getstore($url, $path);
	if(is_error($status)){
		print STDERR "Download failed [$status] ($url)\n";
		return;
	}
	return $path;
}

sub diffFile{
	my $file = shift;
	my $fullPath = fullPath($file);
  my $url = _getUrl($file);
	my ($fh_tmp, $f_tmp) = tempfile();
	downloadFile($url, $f_tmp);
	my $cmd = join(" ", DIFF, DIFF_OPTS, $f_tmp, $fullPath);
	my @lines = `$cmd`;
	return @lines;
}

sub cleanFile{
#	my $dir = shift;
	my $file = shift;
	my $sum = shift;
	my $url = _getUrl($file);
	my $fullPath = fullPath($file);
	downloadFile($url, $fullPath);
	if(fileHasChanged($fullPath, $sum)){
		_warning("\nERROR: Attempted to replace '$fullPath'; failed $!");
		return;
	}
	return $sum;
}

sub sqlShellCmd{
	my $vars = parseVariables(fullPath("wp-config.php"));
	my $cmd = "mysql -u $vars->{'DB_USER'} -p$vars->{'DB_PASSWORD'} -h$vars->{'DB_HOST'} $vars->{'DB_NAME'}";
	return $cmd;
}


sub parseVariables{
	my $file = shift;
	open(my $fh, "<", $file) or die "Failed to open config.php : $!";
	my $vars;
	while(my $line = readline($fh)){
		#define lines:
		if($line =~ /^\s*define\s*\(["']([^"']+)["']\s*,\s*["']([^"']+)["']\s*\)\s*;\s*$/){
			$vars->{$1} = $2;
		# variable assignment:
		}elsif($line =~ /^\s*\$(\S+)\s*=\s*('|")?([\d\.]+)('|")?;\s*$/g){
			$vars->{$1} = $3;
		}
	}
	return $vars;
}

sub getFixPermsCommands{
	my $user = shift;
	my $wpuploads = $dir."/wp-content/uploads";
	my $cgibin = $dir."/cgi-bin";
	$wpuploads =~ s#//#/#g;
	$cgibin =~ s#//#/#g;
	$dir =~ s#//#/#g;

	my @cmds;
	push(@cmds, "chown $user:www-data $dir -R");
	push(@cmds, "find $dir -type f -exec chmod 740 {} ".'\;');
	push(@cmds, "find $dir -type d -exec chmod 750 {} ".'\;');
	push(@cmds, "chmod g+rws,u+s $wpuploads -R");
	if( -d $cgibin ){
		push(@cmds, "chown $user:$user $cgibin -R");
		push(@cmds, "chmod 755 $cgibin -R");
	}
	return @cmds;
}

sub guessUnixUser{
	my $uid = (stat($dir))[4];
	if($uid =~ /\d+/){
		if(my $uname = getpwuid($uid)){
			return $uname;
		}
		return $uid;
	}
	return (split(m#\/#, $dir))[2];
}


sub _error{
	my $message = shift;
	print STDERR $message, "\n";
	exit 1;
} 

sub _warning{
	my $message = shift;
	print STDERR $message, "\n";
}

sub version{
	print $VERSION, "\n";
}
				1
